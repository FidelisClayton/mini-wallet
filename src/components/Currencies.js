import React from 'react'

import Currency from './Currency'

const Currencies = (props) => {
  return (
    <div className="home__currencies">
      <Currency
        name="Real"
        total={100000}
        exchangeable={false}
      />
      <Currency
        name="Brita"
        price={3.73}
        total={67.14}
        amount={18}
        currency="BRT"
      />
      <Currency
        name="Bitcoin"
        price={28000}
        total={250}
        amount={0.00892857}
        currency="BTC"
      />
    </div>
  )
}

export default Currencies
