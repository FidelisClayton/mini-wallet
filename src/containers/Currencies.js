import React, { Component } from 'react'
import { connect } from 'react-redux'

import Currency from '../components/Currency'
import ExchangeForm from '../components/ExchangeForm'

import {
  openModal,
  closeModal
} from '../store/actions/modal'

import { createTransaction } from '../store/actions/transactions'
import { fetchUser } from '../store/actions/user'

export class Currencies extends Component {
  makeTransaction = userId => transaction => {
    return this.props.createTransaction(transaction, userId)
      .then(() => this.props.fetchUser(userId))
  }

  handleSell = (wallet, prices) => {
    this.props.openModal({
      children: (
        <ExchangeForm
          wallet={wallet}
          prices={prices}
          onSell={this.makeTransaction(this.props.user._id)}
          closeModal={this.props.closeModal}
          type="sell"
        />
      ),
      title: `Vender ${wallet.name}`
    })
  }

  handleBuy = (wallet, prices) => {
    const realWallet = this.props.wallets.find(wallet => wallet.id === 'brl')

    this.props.openModal({
      children: (
        <ExchangeForm
          wallet={wallet}
          prices={prices}
          realWallet={realWallet}
          onBuy={this.makeTransaction(this.props.user._id)}
          closeModal={this.props.closeModal}
          type="buy"
        />
      ),
      title: `Comprar ${wallet.name}`
    })
  }

  render () {
    return (
      <div className="home__currencies">
        { this.props.wallets.map(wallet => {
          const prices = this.props.prices.tokens[wallet.token.toLowerCase()]
          const total = wallet.balance * prices.sell

          return (
            <Currency
              key={wallet.token}
              name={wallet.name}
              prices={prices}
              total={total}
              amount={wallet.balance}
              exchangeable={wallet.exchangeable}
              currency={wallet.token}
              onSell={() => this.handleSell(wallet, prices)}
              onBuy={() => this.handleBuy(wallet, prices)}
            />
          )
        })}
      </div>
    )
  }
}

const mapStateToProps = ({
  auth,
  prices
}) => ({
  user: auth.user,
  wallets: auth.user.wallets,
  prices
})

export default connect(
  mapStateToProps,
  {
    openModal,
    closeModal,
    createTransaction,
    fetchUser
  }
)(Currencies)
