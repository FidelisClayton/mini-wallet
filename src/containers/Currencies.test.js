import React from 'react'
import { shallow } from 'enzyme'

import { Currencies } from './Currencies'

const wallets = [
  {
    token: 'BRL',
    name: 'Real',
    balance: 0,
    exchangeable: true,
    currency: 'brl'
  }
]

const prices = {
  tokens: {
    brl: {
      buy: 0,
      sell: 0
    }
  }
}

describe('Component: <Currencies />', () => {
  it('should render properly', () => {
    const wrapper = shallow(
      <Currencies
        wallets={wallets}
        prices={prices}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render currencies', () => {
    const wrapper = shallow(
      <Currencies
        wallets={wallets}
        prices={prices}
      />
    )

    expect(wrapper.find('Currency')).toHaveLength(1)
  })

  it('should open modal when handleSell is invoked', () => {
    const openModal = jest.fn()

    const wrapper = shallow(
      <Currencies
        wallets={wallets}
        prices={prices}
        openModal={openModal}
        user={{ _id: 'user@provider.com' }}
      />
    )

    const instance = wrapper.instance()

    instance.handleSell({
      name: 'Real',
      balance: 0,
      decimalPlaces: 2,
      token: 'brl'
    })()

    expect(openModal.mock.calls.length).toEqual(1)
  })

  it('should open modal when handleBuy is invoked', () => {
    const openModal = jest.fn()

    const wrapper = shallow(
      <Currencies
        wallets={wallets}
        prices={prices}
        openModal={openModal}
        user={{ _id: 'user@provider.com' }}
      />
    )

    const instance = wrapper.instance()

    instance.handleBuy({
      name: 'Real',
      balance: 0,
      decimalPlaces: 2,
      token: 'brl'
    })()

    expect(openModal.mock.calls.length).toEqual(1)
  })

  it('should handle makeTransaction', () => {
    const createTransaction = jest.fn(() => new Promise(resolve => resolve()))
    const fetchUser = jest.fn()

    const wrapper = shallow(
      <Currencies
        wallets={wallets}
        prices={prices}
        createTransaction={createTransaction}
        fetchUser={fetchUser}
        user={{ _id: 'user@provider.com' }}
      />
    )

    const instance = wrapper.instance()

    instance.makeTransaction('userId')({})

    expect(createTransaction.mock.calls.length).toEqual(1)
  })
})
