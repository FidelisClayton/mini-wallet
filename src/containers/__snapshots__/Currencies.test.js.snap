// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Currencies /> should render properly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Currencies
    prices={
        Object {
            "tokens": Object {
              "brl": Object {
                "buy": 0,
                "sell": 0,
              },
            },
          }
    }
    wallets={
        Array [
            Object {
              "balance": 0,
              "currency": "brl",
              "exchangeable": true,
              "name": "Real",
              "token": "BRL",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Currency
          amount={0}
          currency="BRL"
          exchangeable={true}
          name="Real"
          onBuy={[Function]}
          onSell={[Function]}
          prices={
                    Object {
                              "buy": 0,
                              "sell": 0,
                            }
          }
          total={0}
/>,
      ],
      "className": "home__currencies css-102045u",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "BRL",
        "nodeType": "function",
        "props": Object {
          "amount": 0,
          "currency": "BRL",
          "exchangeable": true,
          "name": "Real",
          "onBuy": [Function],
          "onSell": [Function],
          "prices": Object {
            "buy": 0,
            "sell": 0,
          },
          "total": 0,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Currency
            amount={0}
            currency="BRL"
            exchangeable={true}
            name="Real"
            onBuy={[Function]}
            onSell={[Function]}
            prices={
                        Object {
                                    "buy": 0,
                                    "sell": 0,
                                  }
            }
            total={0}
/>,
        ],
        "className": "home__currencies css-102045u",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "BRL",
          "nodeType": "function",
          "props": Object {
            "amount": 0,
            "currency": "BRL",
            "exchangeable": true,
            "name": "Real",
            "onBuy": [Function],
            "onSell": [Function],
            "prices": Object {
              "buy": 0,
              "sell": 0,
            },
            "total": 0,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
